feat: support multiple paths in V2 functions

eduardoboucas opened the pull request:
#### Summary

Allow an array of strings in the `path` in-source configuration property of V2 functions.

The types [already allow this](https://github.com/netlify/functions/blob/3270f5f7823aef8f27e62ab39322514481b7b171/src/function/v2.ts#L10), as well as the rest of the infrastructure.

lukasholzer commented on tests/unit/runtimes/node/in_source_config.test.ts:
```diff
@@ -262,7 +262,7 @@ describe('V2 API', () => {
 
     test('With an invalid pattern', () => {
-      expect.assertions(4)
+      expect.assertions(8)
```
oof 🫣

eduardoboucas replied:
I didn't think it was worth splitting into another test.

eduardoboucas replied:
I've split this into two tests now.

lukasholzer commented on tests/unit/runtimes/node/in_source_config.test.ts:
```diff
@@ -317,5 +336,26 @@ describe('V2 API', () => {
+      }
+  
+      export const config = {
+        path: ["/store/:category/products/:product-id", "/super-awesome-campaign"]
```
can we make the second one a named group as well?

+ Can we add a test for what if both routes are the same?
+ What if I add falsy values like `null` or `undefined` to the array? Do we throw type errors if numbers or whatever is provided? (You have a non nullable function but no test cases for it)

eduardoboucas replied:
I've added more tests to capture this.

lukasholzer commented on src/utils/routes.ts:
```diff
@@ -52,11 +49,22 @@ export const getRoutesFromPath = (path: unknown, functionName: string, methods:
+  const paths = Array.isArray(input) ? input : [input]
+  const routes = paths.map((path) => getRoute(path, functionName, methods ?? [])).filter(nonNullable)
+
+  return routes
```
I would do `return [...new Set(routes)]` to remove duplicates

eduardoboucas replied:
Done in https://github.com/netlify/zip-it-and-ship-it/pull/1586/commits/9e2cac92e2a50eb82f1ae5f112651b3933522d81.

eduardoboucas replied:
We can't do it on `routes` because it's an array of objects. I've done it in `paths`.

lukasholzer commented on src/utils/routes.ts:
```diff
@@ -52,11 +49,22 @@ export const getRoutesFromPath = (path: unknown, functionName: string, methods:
   }
 }
+
+export const getRoutes = (input: unknown, functionName: string, methods: string[]): Route[] => {
```
NIT: would love a small jsdoc that explains the parameters and the method. So that a tipsy on call engineer at 2am in the morning can understand it as well :D

eduardoboucas replied:
Done.

lukasholzer commented on src/utils/routes.ts:
```diff
@@ -17,11 +18,7 @@ const isPathLiteral = (path: string) => {
-    return []
-  }
-
+const getRoute = (path: unknown, functionName: string, methods: string[]): Route | undefined => {
```
NIT: small jsdoc would be nice

eduardoboucas replied:
Done.

lukasholzer reviewed (approved):
Some small comments and request for tests but nothing blocking ;)

eduardoboucas commented:
@lukasholzer I think I addressed all your feedback. I ended up having to make some changes in the AST parsing, because when processing arrays we were only looking for strings, so any other primitive values were ignored (so we couldn't throw a type error if people tried to use them).

This has now been addressed.

lukasholzer reviewed (approved):
Thanks for the added test cases! 🚀
